portal/portal-impl/impl/src/java/org/sakaiproject/portal/charon/handlers/FavoritesHandler.java

```diff
@@ -105,10 +188,44 @@ private void setFavoriteSiteIds(String userId, String[] siteIds) throws Permissi
 		// Replace all existing values
 		props.removeProperty(FAVORITES_PROPERTY);
 
-		for (String siteId : siteIds) {
+		for (String siteId : favorites.favoriteSiteIds) {
 			props.addPropertyToList(FAVORITES_PROPERTY, siteId);
 		}
 
+		props.removeProperty(AUTO_FAVORITE_ENABLED_PROPERTY);
+		props.addProperty(AUTO_FAVORITE_ENABLED_PROPERTY, String.valueOf(favorites.autoFavoritesEnabled));
+
 		PreferencesService.commit(edit);
 	}
+
+	public static class UserFavorites {
+		public List<String> favoriteSiteIds;
+		public boolean autoFavoritesEnabled;
+
+		public UserFavorites() {
+			favoriteSiteIds = Collections.<String>emptyList();
+			autoFavoritesEnabled = false;
+		}
+
+		public String toJSON() {
+			JSONObject obj = new JSONObject();
+
+			obj.put("autoFavoritesEnabled", autoFavoritesEnabled);
+			obj.put("favoriteSiteIds", favoriteSiteIds);
+
+			return obj.toString();
+		}
+
+		public static UserFavorites fromJSON(String json) throws ParseException {
+			JSONParser parser = new JSONParser();
+
+			JSONObject obj = (JSONObject)parser.parse(json);
```

Comment:
How will some of the other exceptions that may occur be handled? That is what if a ClassCastException occurs (say if a JSONArray is returned instead of a JSONObject.